Remove user id from server session

rogersteblerbsi opened the pull request:
As part of the move toward a stateless server session, the `userId` has been removed from the server session.

To get the user id for the current run context, the ThreadLocal `UserId.CURRENT` can be used instead.

423389

rogersteblerbsi committed:
Remove user id from server session

As part of the move toward a stateless server session, the userId has been removed from the server session.

To get the user id for the current run context, the ThreadLocal UserId.CURRENT can be used instead.

423389

rogersteblerbsi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/AbstractServerSession.java:
```diff
@@ -92,7 +87,7 @@ public boolean isStopping() {
   @Override
   public final String getUserId() {
-    return m_userId;
+    return Users.CURRENT.get();
```
Should we completely remove this method from the server session? Since the session is stateless this method does not return the userId of the user, the session belongs to, but the userId of the current run context. So this method may be confusing.

paolobazzi replied:
I'll would move getUserId() to IClientSession and change all code accessing ISession.getUserId to Users.CURRENT.get().

rogersteblerbsi replied:
`getUserId` moved to `IClientSession`. All usages in shared and server replaced by `Users.CURRENT.get()`

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/user/Users.java:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  /**
+   * userId which is currently associated to the current {@link RunContext}.
```
Change javadoc simliar to org.eclipse.scout.rt.shared.ISession#CURRENT?
```suggestion
   * The {@code userId} which is currently associated with the current thread.  
```

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/user/Users.java:
```diff
@@ -0,0 +1,23 @@
+
+import org.eclipse.scout.rt.platform.context.RunContext;
+
+public final class Users {
```
Naming: "Users" vs. "User" - other Scout thread locals use singular names, like ISession.CURRENT, UserAgent,CURRENT etc,

rogersteblerbsi replied:
Current new naming: `UserId.CURRENT`

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/context/ServerRunContext.java:
```diff
@@ -70,7 +72,12 @@ protected <RESULT> void interceptCallableChain(final CallableChain<RESULT> calla
         .add(new ThreadLocalProcessor<>(IClientNodeId.CURRENT, m_clientNodeId))
-        .add(new ThreadLocalProcessor<>(ClientNotificationCollector.CURRENT, m_clientNotificationCollector));
+        .add(new ThreadLocalProcessor<>(ClientNotificationCollector.CURRENT, m_clientNotificationCollector))
+        .add(new ThreadLocalProcessor<>(Users.CURRENT, getUserId()));
```
Implement like m_userAgent as member variable and with getter/setter?

Implement userId in
- org.eclipse.scout.rt.server.context.ServerRunContext#copyValues
- org.eclipse.scout.rt.server.context.ServerRunContext#fillCurrentValues?
- org.eclipse.scout.rt.server.context.ServerRunContext#interceptToStringBuilder

Inital value should then be provided once by org.eclipse.scout.rt.server.context.HttpServerRunContextProducer instead of resolving value every time a server run context is produced?
See org.eclipse.scout.rt.server.commons.context.HttpRunContextProducer#produce(jakarta.servlet.http.HttpServletRequest, jakarta.servlet.http.HttpServletResponse, org.eclipse.scout.rt.platform.context.RunContext)
-> resolve userId once by using the subject

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/context/ClientRunContext.java:
```diff
@@ -68,7 +69,8 @@ protected <RESULT> void interceptCallableChain(final CallableChain<RESULT> calla
         .add(new ThreadLocalProcessor<>(IOutline.CURRENT, m_outline))
-        .add(new ThreadLocalProcessor<>(IForm.CURRENT, m_form));
+        .add(new ThreadLocalProcessor<>(IForm.CURRENT, m_form))
+        .add(new ThreadLocalProcessor<>(Users.CURRENT, m_session != null ? m_session.getUserId() : null));
```
Use member variable to save the value.

And also implement
- copyValues
- fillCurrentValues
- interceptToStringBuilder

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/ServerRunContextProducerTest.java:
```diff
@@ -32,8 +33,7 @@ public void testProduceWithNoSession() {
     ServerRunContext context = producer.produce(getSubjectForPrincipalName(username));
-    IServerSession session = context.getSession();
-    assertEquals(username, session.getUserId());
+    assertEquals(username, context.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/ServerRunContextProducerTest.java:
```diff
@@ -47,7 +47,7 @@ public void testProduceWithSessionDifferentUsername() {
       IServerSession session = context.getSession();
       assertNotEquals(previousSession, session);
-      assertEquals(secondUsername, session.getUserId());
+      assertEquals(secondUsername, context.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/ServerRunContextProducerTest.java:
```diff
@@ -62,16 +62,15 @@ public void testProduceWithSessionSameUsername() {
       IServerSession session = context.getSession();
       assertEquals(previousSession, session);
-      assertEquals(secondUsername, session.getUserId());
+      assertEquals(secondUsername, context.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/ServerRunContextProducerTest.java:
```diff
@@ -62,16 +62,15 @@ public void testProduceWithSessionSameUsername() {
     ServerRunContext context = producer.produce(getSubjectForPrincipalName(principalName));
-    IServerSession session = context.getSession();
-    assertEquals(principalName, session.getUserId());
+    assertEquals(principalName, context.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/session/ServerSessionProviderWithCacheTest.java:
```diff
@@ -102,44 +104,49 @@ public void testNewSessionCacheKey() {
     assertFalse(session.isStopping());
     assertFalse(session.isStopped());
-    assertEquals("anna", session.getUserId());
+    assertEquals("anna", runContext.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/session/ServerSessionProviderWithCacheTest.java:
```diff
@@ -102,44 +104,49 @@ public void testNewSessionCacheKey() {
     assertFalse(session.isStopping());
     assertFalse(session.isStopped());
-    assertEquals("anna", session.getUserId());
+    assertEquals("anna", runContext.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/session/ServerSessionProviderWithCacheTest.java:
```diff
@@ -149,15 +156,16 @@ public void testCacheEvictionStopsSession() {
     assertFalse(otherSession.isStopping());
     assertFalse(otherSession.isStopped());
-    assertEquals("anna", otherSession.getUserId());
+    assertEquals("anna", otherRunContext.call(Users.CURRENT::get));
```
change to context.getUserId() after introducing getter?

rogersteblerbsi replied:
As discussed, we don't add a member variable. We just set it as ThreadLocal.

paolobazzi commented on org.eclipse.scout.rt.server/src/main/java/org/eclipse/scout/rt/server/context/ServerRunContextProducer.java:
```diff
@@ -43,6 +44,7 @@ public class ServerRunContextProducer extends RunContextProducer {
   public ServerRunContext produce(final Subject subject) {
     final ServerRunContext serverRunContext = ServerRunContexts.copyCurrent(true)
         .withSubject(subject)
+        .withThreadLocal(Users.CURRENT, BEANS.get(IAccessControlService.class).getUserId(subject))
```
In client the user is added in org.eclipse.scout.rt.client.context.ClientRunContext#interceptCallableChain
Check client vs. server and use same location to add user thread local.

Preferrable in XyRunContext#interceptCallableChain
And maybe remove then in org.eclipse.scout.rt.server.commons.context.HttpRunContextProducer#produce(jakarta.servlet.http.HttpServletRequest, jakarta.servlet.http.HttpServletResponse, org.eclipse.scout.rt.platform.context.RunContext)

rogersteblerbsi replied:
Setting the UserId ThreadLocal should only be done once in an outer run context and then be copied into the inner ones. So it is done in the producer. When it's set in `ServerRunContext#interceptCallableChain` the value is applied on every call, which may be expensive.

For the ClientRunContext we just apply the already set UserId from the ClientSession, which is not really expensive and also works, when the client session is changed by `withSession`. So there I think it's ok when it's done in the `ClientRunContext#interceptCallableChain`. In future, when the server session will disappear completely, this will probably be asymetric anyways when the `ServerRunContext` will not have a session anymore, but the `ClientRunContext` still has it.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/testing/server/runner/ServerTestRunnerDifferentSessionTest.java:
```diff
@@ -1,5 +1,5 @@
+ * Copyright (c) 2010, 2025 BSI Business Systems Integration AG
  *
  * This program and the accompanying materials are made
  * available under the terms of the Eclipse Public License 2.0
```
add assertion for RunContext.CURRENT.get(User.CURRENT::get)?

rogersteblerbsi replied:
Added assertions for `UserId.CURRENT.get()`

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/testing/server/runner/ServerTestRunnerDifferentSubjectTest.java:
```diff
@@ -39,7 +39,6 @@ public static void beforeClass() {
     m_serverSessions = new HashSet<>();
     ISession serverSession = IServerSession.CURRENT.get();
     assertTrue(serverSession instanceof JUnitServerSession);
```
add assertion for RunContext.CURRENT.get(User.CURRENT::get)

rogersteblerbsi replied:
Added assertions for `UserId.CURRENT.get()`

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/testing/server/runner/ServerTestRunnerSameSessionTest.java:
```diff
@@ -1,5 +1,5 @@
+ * Copyright (c) 2010, 2025 BSI Business Systems Integration AG
  *
  * This program and the accompanying materials are made
  * available under the terms of the Eclipse Public License 2.0
```
add assertion for RunContext.CURRENT.get(User.CURRENT::get)

rogersteblerbsi replied:
Added assertions for `UserId.CURRENT.get()`

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/session/Sessions.java:
```diff
@@ -61,18 +60,10 @@ public static String randomSessionId() {
-      return session.getUserId();
-    }
-    return BEANS.get(IAccessControlService.class).getUserIdOfCurrentSubject();
+    return Users.CURRENT.get();
```
do we need this method in future?

rogersteblerbsi replied:
Method removed. Callers just check for `UserId.CURRENT.get()` instead

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/context/ClientRunContext.java:
```diff
@@ -68,7 +69,8 @@ protected <RESULT> void interceptCallableChain(final CallableChain<RESULT> calla
         .add(new ThreadLocalProcessor<>(IOutline.CURRENT, m_outline))
-        .add(new ThreadLocalProcessor<>(IForm.CURRENT, m_form));
+        .add(new ThreadLocalProcessor<>(IForm.CURRENT, m_form))
+        .add(new ThreadLocalProcessor<>(Users.CURRENT, m_session != null ? m_session.getUserId() : null));
```
Move line directly after session thread local?

rogersteblerbsi replied:
Moved

rogersteblerbsi committed:
Add ThreadLocal and HTTP header for SessionId

As part of the move toward a stateless server session, a TheadLocal has
been introduced with the session id of the client session.

The clientId is sent to the backend via custom HTTP header X-Scout-Client-Session-Id

423389

paolobazzi commented on eclipse-scout-core/src/App.ts:
```diff
@@ -502,8 +502,9 @@ export class App extends EventEmitter {
     request.setRequestHeader('X-Scout-Correlation-Id', numbers.correlationId());
     request.setRequestHeader('X-Requested-With', 'XMLHttpRequest'); // explicitly add here because jQuery only adds it automatically if it is no crossDomain request
-    if (this.sessions[0] && this.sessions[0].ready) {
+    if (this.sessions[0]?.ready) {
```
Check this change with @bschwarzent

rogersteblerbsi replied:
Removed session id header again in requests from UI/browser for now. User could manipulate it and depending on what is done with this id it could be a security issue.
In requests from Client to Server however, the header is still set.

paolobazzi commented on org.eclipse.scout.rt.server.jdbc/src/main/java/org/eclipse/scout/rt/server/jdbc/internal/exec/StatementProcessor.java:
```diff
@@ -114,25 +112,11 @@ public StatementProcessor(ISqlService callerService, String stm, Object[] bindBa
-          m_bindBases[m_bindBases.length - 1] = session;
-        }
+      if (bindBases == null) {
+        m_bindBases = new Object[]{};
```
as replacement, we could add just the current userId here as default bind

rogersteblerbsi replied:
`userId` is now added as default bind

paolobazzi commented on org.eclipse.scout.rt.server.commons.test/src/test/java/org/eclipse/scout/rt/server/commons/context/HttpRunContextProducerTest.java:
```diff
@@ -40,6 +46,23 @@ public class HttpRunContextProducerTest {
   static final Locale TEST_LOCALE = Locale.CANADA_FRENCH;
   static final Subject TEST_SUBJECT = BEANS.get(ServletFilterHelper.class).createSubject(BEANS.get(SimplePrincipalProducer.class).produce(TEST_SUBJECT_NAME));
 
+  protected static IBean<?> s_accessControlMockBean;
```
- change to non-static and use `@BeanMock`
- remove cleanup method
- change setup to just contain `when(accessControlMock.getUserIdOfCurrentSubject()).thenReturn(TEST_SUBJECT_NAME);`

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/logging/UserIdContextValueProvider.java:
update javadoc

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/session/SessionId.java:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  /**
+   * The {@code session id} which is currently associated with the current thread.
```
I'll would mention that this is the id of the **client** session

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/user/UserId.java:
```diff
@@ -0,0 +1,21 @@
+  }
+
+  /**
+   * The {@code userId} which is currently associated with the current thread.
```
```suggestion
   * The {@code userId} which is currently associated with the current thread. The {@code userId} is extracted from current
   * {@link Subject} using {@link IAccessControlService#getUserIdOfCurrentSubject()} and usually corresponds to the username the of logged-in user.
```

paolobazzi commented on org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/rest/SessionIdClientRequestFilter.java:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  protected void addSessionIdHeader(ClientRequestContext requestContext) {
+    if (requestContext.getHeaders().get(CLIENT_SESSION_ID_HTTP_HEADER) == null) {
```
which case leads to an already set session id header?

rogersteblerbsi replied:
I think this should not happen. Removed. Was copied from the `IdSignatureClientRequestFilter`.

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/servicetunnel/ServiceTunnelServiceTest.java:
```diff
@@ -189,6 +191,16 @@ public void testIncomingRequest() throws Exception {
+      ServiceTunnelService s = BEANS.get(ServiceTunnelService.class);
+      ServiceTunnelRequest serviceRequest = prepareTestRequest();
+      ServerRunContext context = s.createServiceTunnelRunContext(serviceRequest);
+      assertEquals("testId", context.call(() -> SessionId.CURRENT.get()));
```
```suggestion
      assertEquals("testId", context.call(SessionId.CURRENT::get));
```

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/HttpServerRunContextProducerTest.java:
```diff
@@ -52,14 +53,16 @@ public void testGetOrCreateScoutSessionWithRandomId() {
     try {
       assertNotNull(session);
       assertNotNull(session.getId());
+      assertEquals(clientSessionId, context.call(() -> SessionId.CURRENT.get()));
```
change to SessionId.CURRENT::get

paolobazzi commented on org.eclipse.scout.rt.server.test/src/test/java/org/eclipse/scout/rt/server/context/HttpServerRunContextProducerTest.java:
```diff
@@ -71,29 +74,32 @@ public void testGetOrCreateScoutSessionWithExistingSessionIdOnHttpSession() {
       assertNotNull(session);
-      assertEquals(existingScoutSessionId, session.getId());
+      assertEquals(existingServerSessionId, session.getId());
+      assertEquals(clientSessionId, context.call(() -> SessionId.CURRENT.get()));
```
change to SessionId.CURRENT::get

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/session/SessionId.java:
```diff
@@ -0,0 +1,26 @@
+  /**
+   * Name of the HTTP header to transport the {@code session id}.
+   */
+  public static final String CLIENT_SESSION_ID_HTTP_HEADER = "X-Scout-Client-Session-Id";
```
rename to `HTTP_HEADER_NAME` like org.eclipse.scout.rt.platform.context.CorrelationId#HTTP_HEADER_NAME and org.eclipse.scout.rt.shared.authentication.DefaultAuthToken#HTTP_HEADER_NAME

paolobazzi commented on org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/servicetunnel/ServiceTunnelRequest.java:
remove unused field

paolobazzi reviewed (approved):
Looks good!
